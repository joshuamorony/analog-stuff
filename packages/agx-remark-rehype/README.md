# agx-remark-rehype

This library was generated with [Nx](https://nx.dev).

## Building

Run `nx build agx-remark-rehype` to build the library.

## Running unit tests

Run `nx test agx-remark-rehype` to execute the unit tests via [Vitest](https://vitest.dev/).
