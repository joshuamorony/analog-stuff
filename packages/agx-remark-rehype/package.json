{
  "name": "@joshmorony/agx-remark-rehype",
  "version": "0.0.9",
  "repository": {
    "type": "git",
    "url": "https://github.com/joshuamorony/analog-stuff"
  },
  "description": "Uses unified and remark/rehype ecosystem for parsing agx files in AnalogJS",
  "keywords": [
    "analog",
    "angular",
    "markdown",
    "remark",
    "rehype"
  ],
  "author": {
    "name": "Joshua Morony"
  },
  "homepage": "https://github.com/joshuamorony/analog-stuff/tree/main/packages/agx-remark-rehype",
  "bugs": {
    "url": "https://github.com/joshuamorony/analog-stuff/issues"
  },
  "license": "MIT",
  "dependencies": {
    "rehype-stringify": "^10.0.0",
    "remark-extract-frontmatter": "^3.2.0",
    "remark-frontmatter": "^5.0.0",
    "remark-parse": "^11.0.0",
    "remark-rehype": "^11.1.0",
    "unified": "^11.0.5",
    "unist": "^0.0.1",
    "unist-util-visit": "^5.0.0"
  }
}
