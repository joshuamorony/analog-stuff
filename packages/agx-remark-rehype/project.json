{
  "name": "agx-remark-rehype",
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "sourceRoot": "packages/agx-remark-rehype/src",
  "projectType": "library",
  "tags": [],
  "targets": {
    "test": {
      "executor": "@nx/vite:test",
      "outputs": ["{options.reportsDirectory}"],
      "options": {
        "reportsDirectory": "../../coverage/packages/agx-remark-rehype"
      }
    }
  }
}
