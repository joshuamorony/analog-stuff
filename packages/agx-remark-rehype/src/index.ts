export * from './lib/agx-remark-rehype';
