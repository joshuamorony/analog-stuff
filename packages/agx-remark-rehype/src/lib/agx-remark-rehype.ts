import type { Plugin, Processor } from 'unified';
import { unified } from 'unified';
import { Literal } from 'unist';
import { visit } from 'unist-util-visit';
import rehypeStringify from 'rehype-stringify';
import remarkParse from 'remark-parse';
import remarkRehype from 'remark-rehype';

type MarkdownTemplateTransform = (
  content: string,
  fileName: string
) => string | Promise<string>;

type PluginWithSettings<
  S extends any[] = [Record<string, any>?],
  P extends Plugin<S> = Plugin<S>
> = [P, ...S];

type PluginWithoutSettings<S extends any[] = any[]> = (...settings: S) => any;

export type UnifiedPlugins = Array<PluginWithSettings | PluginWithoutSettings>;

export interface RemarkRehypeOptions {
  remarkPlugins?: UnifiedPlugins;
  rehypePlugins?: UnifiedPlugins;
}

export const agxRemarkRehype =
  (options: RemarkRehypeOptions = {}): MarkdownTemplateTransform =>
  async (content: string) => {
    const processor = setupUnified(options);
    const mdContent = await processor.process(content);

    return fixDoubleEscape(String(mdContent));
  };

const fixDoubleEscape = (content: string) => {
  return content
    .replace(/&#x26;#64;/g, '&#64;')
    .replace(/&#x26;#x2774;/g, '&#x2774;')
    .replace(/&#x26;#x2775;/g, '&#x2775;');
};

const rehypeAnalog: Plugin = () => {
  return (tree) => {
    visit(tree, 'element', (node) => {
      if ((node as any).tagName === 'p') {
        visit(node, 'text', (childNode) => {
          const literal = childNode as Literal;
          if (detectAngularControlFlow(literal.value as string)) {
            literal.type = 'raw';
          } else {
            literal.value = escapeBreakingCharacters(literal.value as string);
          }
        });
      }

      if ((node as any).tagName === 'code') {
        visit(node, 'text', (childNode) => {
          const literal = childNode as Literal;
          literal.value = escapeBreakingCharacters(literal.value as string);
        });
      }
    });
  };
};

const escapeBreakingCharacters = (code: string) => {
  // Escape commonly used HTML characters
  // in Angular templates that cause template parse errors
  // such as @, {, and ,}.
  code = code.replace(/@/g, '&#64;');
  code = code.replace(/{/g, '&#x2774;').replace(/}/g, '&#x2775;');
  return code;
};

const detectAngularControlFlow = (text: string) => {
  return (
    (text.trim().startsWith('@if') ||
      text.trim().startsWith('@for') ||
      text.trim().startsWith('@switch') ||
      text.trim().startsWith('@defer')) &&
    text.trim().endsWith('}')
  );
};

const applyPlugins = (
  plugins: UnifiedPlugins,
  parser: Processor<any, any, any>
) => {
  plugins.forEach((plugin: any) => {
    if (Array.isArray(plugin)) {
      if (plugin[1] && plugin[1]) parser.use(plugin[0], plugin[1]);
      else parser.use(plugin[0]);
    } else {
      parser.use(plugin);
    }
  });

  return parser;
};

const setupUnified = (options: RemarkRehypeOptions) => {
  const toMDAST = unified().use(remarkParse);
  applyPlugins(options.remarkPlugins || [], toMDAST);
  const toHAST = toMDAST.use(remarkRehype, { allowDangerousHtml: true });
  applyPlugins(options.rehypePlugins || [], toHAST);

  const processor = toHAST
    .use(rehypeStringify, {
      allowDangerousHtml: true,
    })
    .use(rehypeAnalog);

  return processor;
};
