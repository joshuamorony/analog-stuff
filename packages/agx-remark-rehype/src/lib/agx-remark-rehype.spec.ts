import { agxRemarkRehype } from './agx-remark-rehype';

describe('agxRemarkRehype', () => {
  it('should return standard markdown unmodified', async () => {
    const transform = agxRemarkRehype();
    const content = await transform(standardMarkdown, 'test.agx');
    expect(content).toMatchSnapshot();
  });

  it('should escape characters that break Angular templates', async () => {
    const transform = agxRemarkRehype();
    const content = await transform(breakingCharactersMarkdown, 'test.agx');
    expect(content).toMatchSnapshot();
  });

  it('should allow embedding components without escaping', () => {});

  it('should allow Angular control flow syntax without escaping', () => {});

  it('should allow supplying remark plugins', () => {});

  it('should allow supplying rehype plugins', () => {});
});

const standardMarkdown = `
## Test
Hello
* one
* two
* three
`;

const breakingCharactersMarkdown = `
## Test
Hello
Watch out for @ characters, they upset Angular. A { can be quite scary too.
\`\`\`ts
{}
@
\`\`\`
`;
