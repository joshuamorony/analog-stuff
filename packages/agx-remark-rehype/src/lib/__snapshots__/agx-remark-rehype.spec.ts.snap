// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`agxRemarkRehype > should escape characters that break Angular templates 1`] = `
"<h2>Test</h2>
<p>Hello
Watch out for &#64; characters, they upset Angular. A &#x2774; can be quite scary too.</p>
<pre><code class="language-ts">&#x2774;&#x2775;
&#64;
</code></pre>"
`;

exports[`agxRemarkRehype > should return standard markdown unmodified 1`] = `
"<h2>Test</h2>
<p>Hello</p>
<ul>
<li>one</li>
<li>two</li>
<li>three</li>
</ul>"
`;
